hammer/vlsi/hammer_vlsi_impl.py

```diff
@@ -2059,11 +2059,52 @@ def sdf_file(self, value: Optional[str]) -> None:
     ### END Generated interface HammerTimingTool ###
 
 class HasUPFSupport(HammerTool):
-    """Mix-in trait with functions useful for tools with UPF style power
-    constraints"""
-    @property
-    def upf_power_specification(self) -> str:
-        raise NotImplementedError("Automatic generation of UPF power specifications is not supported yet.")
+   """Mix-in trait with functions useful for tools with UPF style power constraints"""
+   @property
+   def upf_power_specification(self) -> str:
+        #Constants and objects
+        output = [] # type: List[str]
+        domain = "AO"
+        #Header
+        output.append("set_design_top {t}".format(t=self.top_module))
+        vdd = VoltageValue(self.get_setting("vlsi.inputs.supplies.VDD"))
+        #Create Single Power Domain
+        output.append("create_power_domain {d} \\".format(d=domain))
+        output.append("\t-elements {.}")
+        #Get Supply Nets
+        power_nets = self.get_all_power_nets() 
+        ground_nets = self.get_all_ground_nets()
+        #Create Supply Ports 
+        for pg_net in (power_nets+ground_nets):
+            if(pg_net.pin != None):
+                #Create Supply Nets
+                output.append("create_supply_net {p} -domain {d}".format(p=pg_net.name, d=domain))
+                output.append("create_supply_port {p} -domain {d} \\".format(p=pg_net.name, d=domain))
+                output.append("\t-direction in")
+                #Connect Supply Net
+                output.append("connect_supply_net {p} -ports {p}".format(p=pg_net.name))
+        #Set Domain Supply Net
+        output.append("set_domain_supply_net {d} \\".format(d=domain))
+        output.append("\t-primary_power_net {p} \\".format(p=power_nets[0].name))
+        output.append("\t-primary_ground_net {g}".format(g=ground_nets[0].name))
+        #Add Port States 
+        for p_net in power_nets:
+            if(p_net.pin != None):
+                output.append("add_port_state {g} \\".format(g=p_net.name))
+                output.append("\t-state {{default {v}}}".format(v=vdd.value))
+        for g_net in ground_nets:
+            if(g_net.pin != None):
+                output.append("add_port_state {g} \\".format(g=g_net.name))
+                output.append("\t-state {default off}")
```

Comment:
I don't think this is correct according to my understanding of the command... instead it should be a voltage of 0. "off" is distinctly a different state?

Reply:
<img width="877" alt="Screen Shot 2023-03-30 at 5 49 42 PM" src="https://user-images.githubusercontent.com/96459298/228982198-cd50ee58-71c3-4b48-92e1-267bad5943d8.png">
It appears to be a special exception according to the 2018 manual, but I will change it to be safe.